### PR TITLE
FV Cell-Area Loss Weighting: geometry-aware loss for non-uniform meshes

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current training loss treats every mesh node equally. CFD meshes are deliberately non-uniform: boundary-layer cells are tiny (O(1e-4) chord) while freestream cells are large (O(0.1) chord). When loss is unweighted, a single large freestream cell contributes equal gradient signal as ten boundary-layer cells — biasing training toward coarse freestream regions, away from the surface where MAE is actually measured.

**Fix:** Weight the VOLUME loss by `1/sqrt(cell_area)` — emphasizing small boundary-layer cells where surface physics live. Surface loss is left unweighted (hard-node mining already handles it).

**Theory:** Finite Volume theory says the correct numerical integration weight for a node is its control-volume area. Lam et al. ICML 2024 (arXiv:2402.02367) showed 15-40% MAE improvement on non-uniform mesh surrogates using FV-style loss weighting. UNTRIED in 1903 experiments. Addresses a fundamental structural bias in the loss formulation.

**Key distinction from hard-node mining:** Hard-node mining weights by ERROR magnitude. Cell-area weighting is geometric — targets structural mesh bias, not error magnitude. Orthogonal.

## Instructions

Add `--fv_area_loss_weight` flag. Compute approximate dual-cell area from k-nearest-neighbor distances, weight volume loss by `1/sqrt(approx_area)`.

### Step 1: Compute approximate cell area

```python
from torch_geometric.nn import knn_graph

def compute_area_weights(pos, k=6):
    """
    pos: [B*N, 2] flattened node positions (batch.pos in PyG)
    Returns: [B*N] area weights (1/sqrt(approx_area)), normalized to mean=1
    """
    edge_index = knn_graph(pos, k=k, loop=False)  # [2, B*N*k]
    src, dst = edge_index
    dists = (pos[dst] - pos[src]).norm(dim=-1)     # [B*N*k]
    mean_dist = torch.zeros(pos.shape[0], device=pos.device)
    mean_dist.scatter_reduce_(0, src, dists, reduce='mean', include_self=False)
    mean_dist = mean_dist.clamp(min=1e-8)
    approx_area = mean_dist ** 2
    area_weight = 1.0 / approx_area.sqrt()
    area_weight = area_weight / area_weight.mean()  # normalize: mean=1, loss scale preserved
    return area_weight  # [B*N]
```

### Step 2: Apply to volume loss only (NOT surface loss)

```python
if args.fv_area_loss_weight:
    pos_flat = batch.pos                     # PyG batch: [B*N, 2]
    area_weights = compute_area_weights(pos_flat, k=6)  # [B*N]
    # Reshape to match your vol_loss shape and apply BEFORE .mean() reduction
    # vol_loss_per_node: [B, N] or [B*N] — multiply BEFORE reducing
    vol_loss = vol_loss * area_weights.view_as(vol_loss)
```

**IMPORTANT:** Find exactly where `vol_loss` is computed in `train.py` and apply weights BEFORE the `.mean()` call. Do NOT apply to surface loss (that has hard-node mining).

### Step 3: Add the flag

```python
parser.add_argument('--fv_area_loss_weight', action='store_true',
                    help='Weight vol loss by 1/sqrt(cell area) — FV mesh-aware loss weighting')
```

### Step 4: Sanity check before full training

Print on first batch: `area_weights.min(), area_weights.max(), area_weights.mean()`. Expect min~0.01, max~50-100, mean~1.0. Surface/boundary-layer nodes should have larger weights than freestream nodes. If all weights are similar, something is wrong with the pos scaling.

### Training commands

```bash
# Seed 42
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent nezuko --seed 42 \
  --wandb_name "nezuko/fv-cell-area-s42" \
  --wandb_group "fv-cell-area-loss-weight" \
  --fv_area_loss_weight \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0

# Seed 73: same but CUDA_VISIBLE_DEVICES=1, --seed 73, --wandb_name "nezuko/fv-cell-area-s73"
```

## Baseline

Current baseline (PR #2290, Re-Stratified Sampling, merged 2026-04-08):

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.74** | < 11.74 |
| p_oodc | 7.65 | < 7.65 |
| **p_tan** | **27.90** | < 27.90 |
| p_re | 6.40 | < 6.40 |

**Reference:** Lam et al. "Finite Volume Features, Global Geometry Representations, and Residual Training for Deep Learning-Based CFD Simulation" (ICML 2024, arXiv:2402.02367)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### FV Cell-Area Loss Weighting — 2-seed experiment

| Metric | Baseline (2-seed avg) | FV Area (S42) | FV Area (S73) | FV Area (2-seed avg) | Δ |
|--------|----------------------|---------------|---------------|---------------------|---|
| **p_in** | **11.74** | 11.9 | 11.8 | **11.85** | **+0.9%** ❌ |
| **p_oodc** | **7.65** | 8.2 | 7.6 | **7.90** | **+3.3%** ❌ |
| **p_tan** | **27.90** | 28.6 | 28.9 | **28.75** | **+3.0%** ❌ |
| **p_re** | **6.40** | 6.6 | 6.5 | **6.55** | **+2.3%** ❌ |

### Details

- **S42**: Best epoch 147/148 (180 min timeout), val_loss=0.3934, peak VRAM=46.1GB
- **S73**: Best epoch 148/148 (180 min timeout), val_loss=0.3862, peak VRAM=46.4GB
- **W&B runs**: `e12wkmrf` (S42), `9a9vbbkz` (S73)
- **Group**: `fv-cell-area-loss-weight`

### Training command
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent nezuko --seed 42 \
  --wandb_name "nezuko/fv-cell-area-s42" \
  --wandb_group "fv-cell-area-loss-weight" \
  --fv_area_loss_weight \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0
```

### Implementation details
- Precomputed per-sample area weights at startup using scipy cKDTree (k=6 neighbors), parallelized across 16 workers
- Precompute time: ~120s for 1322 training samples (one-time cost, not per-epoch)
- Area weight = 1/sqrt(mean_knn_dist²), normalized to mean=1.0 per sample
- Weight stats across all training samples: min=0.0003, max=10.88, mean=1.0, std=1.83
- Applied to ALL vol_loss paths (default, boundary_aware, PCGrad 2-way, PCGrad 3-way, SAM)
- Area weights passed through DataLoader via wrapped dataset + custom collate
- No per-batch overhead — zero slowdown compared to baseline

### What happened
**Negative result.** All four metrics regressed — p_oodc (+3.3%) and p_tan (+3.0%) most significantly.

**Why it likely failed:**
1. **The model already handles mesh non-uniformity well.** The existing Transolver architecture with Physics Attention naturally adapts to the mesh via learned kernel weights and position-aware Fourier PE. The FV area weighting adds redundant information that the model can already learn implicitly.
2. **Extreme weight variance hurts optimization.** Area weights range from 0.0003 to 10.88 (std=1.83), meaning some nodes contribute 36000× more to the loss than others. This effectively reduces the "useful" batch size and adds noise to gradients, particularly for rare boundary-layer nodes that already get well-served by the existing boundary_aware and surface_refine mechanisms.
3. **Volume loss is already well-regularized.** The baseline has progressive volume node ramping, PCGrad, and dedicated surface refinement heads — the volume loss is not the bottleneck. Over-weighting dense mesh regions (which tend to be in the boundary layer) may actually be counterproductive since these regions are already handled by the surface loss term.

### Suggested follow-ups
- **Soft area weighting:** Instead of raw 1/sqrt(area), try a much gentler version like `weight = 1 + 0.1*(1/sqrt(area) - 1)` to preserve the signal while limiting the variance.
- **Area weighting on surface loss only:** CFD surface meshes have particularly non-uniform spacing near leading/trailing edges. Applying FV weighting to just the surface loss might help where it matters most.
- **Hybrid approach:** Only apply area weights in the first N epochs as a curriculum, then switch to uniform weighting. This could help early optimization without hurting fine-tuning.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2297 because ADVISOR: Closing — all metrics regressed: p_in +0.9%, p_oodc +3.3%, p_tan +3.0%, p_re +2.3%. Good analysis: (1) Transolver's physics attention + Fourier PE already adapts to mesh non-uniformity, making FV weighting redundant, (2) extreme weight variance (0.0003 to 10.88) effectively reduces useful batch size and adds gradient noise, (3) volume loss already well-regularized by PCGrad + surface refinement + boundary-aware paths. The gentler variant you suggest (soft area weighting) is interesting but the existing mesh-adaptive mechanisms make this direction low-priority. FV area loss weighting added to DO NOT REVISIT. W&B: e12wkmrf (s42), 9a9vbbkz (s73)..
